Remove support for compiler plugins.

nnethercote opened the pull request:
They've been deprecated for four years.

This commit includes the following changes.
- It eliminates the `rustc_plugin_impl` crate.
- It changes the language used for lints in `compiler/rustc_driver_impl/src/lib.rs` and `compiler/rustc_lint/src/context.rs`. External lints are now called "loaded" lints, rather than "plugins" to avoid confusion with the old plugins. This only has a tiny effect on the output of `-W help`.
- E0457 and E0498 are no longer used.
- E0463 is narrowed, now only relating to unfound crates, not plugins.
- The `plugin` feature was moved from "active" to "removed".
- It removes the entire plugins chapter from the unstable book.
- It removes quite a few tests, mostly all of those in `tests/ui-fulldeps/plugin/`.

Closes #29597.

r? @ghost

nnethercote committed:
Remove support for compiler plugins.

They've been deprecated for four years.

This commit includes the following changes.
- It eliminates the `rustc_plugin_impl` crate.
- It changes the language used for lints in
  `compiler/rustc_driver_impl/src/lib.rs` and
  `compiler/rustc_lint/src/context.rs`. External lints are now called
  "loaded" lints, rather than "plugins" to avoid confusion with the old
  plugins. This only has a tiny effect on the output of `-W help`.
- E0457 and E0498 are no longer used.
- E0463 is narrowed, now only relating to unfound crates, not plugins.
- The `plugin` feature was moved from "active" to "removed".
- It removes the entire plugins chapter from the unstable book.
- It removes quite a few tests, mostly all of those in
  `tests/ui-fulldeps/plugin/`.

Closes #29597.

nnethercote commented:
I decided to implement the removal to see what it looked like, and to encourage a decision. An FCP will be needed, at the minimum.

Cargo will need some adjustment, to remove the `plugin = true` option under `[lib]`. `src/doc/book/redirects/compiler-plugins.md` will also need adjustment to avoid some test failures.

lnicola commented on src/tools/rust-analyzer/crates/hir-def/src/attr/builtin.rs:
```diff
@@ -262,24 +262,6 @@ pub const INERT_ATTRIBUTES: &[BuiltinAttribute] = &[
         naked_functions, experimental!(naked)
     ),
 
-    // Plugins:
```
Given that this is a no-op and we're currently [unable](https://github.com/rust-lang/rust/pull/116136#issuecomment-1733307521) to do syncs, would you mind dropping this change and optionally filing it against the RA repo?

nnethercote replied:
Sure, if this PR is approved I can do that.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6402604150/job/17379532600) [(plain)](https://github.com/rust-lang/rust/commit/50793100dbbb9ccc6f780f53a3921b8b8348519f/checks/17379532600/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.11s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
book/compiler-plugins.html:37: broken link - `unstable-book/language-features/plugin.html`
book/compiler-plugins.html:39: broken link - `unstable-book/language-features/plugin.html`
number of HTML files scanned: 34809
number of HTML redirects found: 10548
number of links checked: 3698238
number of links ignored due to external: 92958
```

</details>

fbstj commented on compiler/rustc_driver_impl/src/lib.rs:
```diff
@@ -1075,20 +1070,22 @@ Available lint options:
+        }
+        (false, ..) => panic!("didn't load additional lints but got them anyway!"),
+        (true, 0, 0) => {
+            safe_println!("This crate does not load any additional lints  or lint groups.")
```
```suggestion
            safe_println!("This crate does not load any additional lints or lint groups.")
```

nnethercote replied:
Fixed, thanks.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6403794589/job/17383052065) [(plain)](https://github.com/rust-lang/rust/commit/0fcfd98bfc667040485f3d592cc41d4ee6087ac9/checks/17383052065/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.10s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
book/compiler-plugins.html:37: broken link - `unstable-book/language-features/plugin.html`
book/compiler-plugins.html:39: broken link - `unstable-book/language-features/plugin.html`
number of HTML files scanned: 34809
number of HTML redirects found: 10548
number of links checked: 3698242
number of links ignored due to external: 92958
```

</details>

compiler-errors commented:
I think the PR description justifies the removal, but also see [this zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Plugin.20support.20removal.3F) for justification.

@rfcbot fcp merge

petrochenkov commented:
@rfcbot servo-buy-in

The main (and only?) reason why this feature wasn't removed several years ago is its servo use case.
It needs to be explicitly confirmed that either servo can migrate to something else, or can remove its lints, or is no longer important enough to care about.

petrochenkov commented:
@rfcbot concern servo-buy-in

matthewjasper commented:
previous attempt at this: #64675

bors commented:
:umbrella: The latest upstream changes (presumably #116455) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
> Servo is often given as a reason for keeping plugin support, but @SimonSapin [said](https://github.com/rust-lang/rust/issues/29597#issuecomment-1006748104) this 18 months ago:
> 
> > Activity in the Servo project has been low lately and will be in the forseeable future. If a better API becomes available, it’s not clear when or if we’d be able to migrate the GC rooting lint to it. Given this, I feel it’s probably not fair to expect rustc to indefinitely keep the burden of maintaining plugin support for Servo’s sake.
> 
> Servo has become more active again recently. I don't know if that has any impact on the decision. Four years of deprecation is a long time!

cc some [Servo people](https://github.com/orgs/servo/people) for input: @mrobinson, @emilio, @mukilan, @sagudev, @jdm, @Manishearth

Manishearth commented:
It's still used in servo for GC safety, yes, and people have been actively working on Servo again.

jdm commented:
Yeah, I don't think anything has really changed here from what's been previously articulated in https://github.com/rust-lang/rust/pull/64675. We had a discussion about whether a linter alternative could work in https://servo.zulipchat.com/#narrow/stream/263398-general/topic/Dylint.

nnethercote commented:
Right, so we're in a "nothing has changed in three years" situation. I'm really hoping to avoid having this long-deprecated, long-unmaintained feature kept around indefinitely.

workingjubilee commented:
Servo will have 100% buy-in on putting in the necessary maintenance work, one way or another, if it is removed now and they wish to use a stable version of 1.75, or the nightly immediately after this lands.

Manishearth commented:
I will say, while servo is the only permanent user that I know of, having a permaunstable plugin API like this has been useful at various times and imo is a good thing to have. Similarly, clang has the unstable clang plugin API that a lot of tools use.

I'd be surprised if this were a maintenance burden: to me "it's been around for years and nobody has tried to remove it yet" feels more due to it not getting in the way of anything, but I could be wrong. servo hasn't had major pull in the rust project for ages so I'd be surprised if the reason was that it _did_ get in the way but people decided not to remove it because of servo.


(while I'm still somewhat biased here; I will note that I haven't written code on servo/servo for ages and wouldn't be directly affected by this decision. I'm just a fan of us having such an API. I've found both the rust and clang plugin APIs to be super useful for various operations)

workingjubilee commented:
> I'd be surprised if this were a maintenance burden: to me "it's been around for years and nobody has tried to remove it yet" feels more due to it not getting in the way of anything, but I could be wrong. servo hasn't had major pull in the rust project for ages so I'd be surprised if the reason was that it did get in the way but people decided not to remove it because of servo.

It is actually insanely annoying to have around, and I have complained about it periodically, because the tests are almost always failing on actual local builds on platforms that aren't x86-64 Linux, which tricks contributors into thinking they broke something. It greatly inconvenienced me the entire time I was trying to contribute to Rust while using Windows as my OS. I at some point stopped using Windows, but I have had to **repeatedly** reassure other contributors that no, they didn't break anything. If I had the courage and knowledge to do so, then, I would've opened this very same PR years ago.

So yes, this is *exactly* why. Literally, petrochenkov raised a concern based on this reason.

No one is maintaining this interface, and rustc wrappers like your very own clippy actually _work_ in a cross-platform way, and are capable of more besides.

Manishearth commented:
Okay, thanks, in which case I feel comfortable about having this removed. I think such a feature would be valuable to have in the long run, I agree that it should be maintained and should not be retained if it is at all a current maintenance burden, which it clearly is.

(And future attempts to re-add this can refer to this thread to understand what things need to be resolved for such a feature to be Good)

sagudev commented:
I think servo should no longer be problematic, I made prototype in https://github.com/servo/servo/pull/30508 (with all of our lints ported), now I only need to polish integration with servo. I would be nice if this PR waited for https://github.com/servo/servo/pull/30508 to land in servo (which should be in a week).

lnicola commented:
> It would be nice if this PR waited for https://github.com/servo/servo/pull/30508 to land in servo (which should be in a week).

Given that Servo is still on `nightly-2023-02-01` and https://github.com/servo/servo/actions/workflows/nightly-rust.yml has been failing since mid-February for seemingly unrelated reasons (can't install `taplo`?), waiting an extra week is unlikely to make a difference, though.

workingjubilee commented:
FCP means this is landing no sooner than the 17th anyways.

nnethercote commented:
We now have 13/15 approvals.

@petrochenkov, you registered the servo-buy-in concern but I think that has been addressed above. Are you happy to mark it as resolved?

petrochenkov commented:
@rfcbot resolve servo-buy-in

BartMassey commented:
Many many years ago some of my students spent a bunch of time building a tool for writing formal contracts for Rust code and having them automatically checked by Z3, which was pretty cool. Sadly, this was right before big changes to MIR, so their work effectively died immediately.

That said, this is the kind of thing that compiler plugins might support nicely.

As not-a-compiler-maintainer I get that the infrastructure may be a pain: I wish there was a better way than just deleting it and never looking back, though I can't think of one offhand.

bjorn3 commented:
The driver interface (rustc_driver) should work for that too, right? There are already several verification tools like Kani that use the driver interface. That interface is not going anywhere.

bjorn3 commented:
Also the current plugin interface only supports lints anymore anyway. It used to support macros too before proc macro got introduced. It has never supported overriding queries or running right before codegen. For that you have always needed the driver interface.

nnethercote commented:
r? @bjorn3

There are some pending conflicts, but they're all trivial: in removed files, or `Cargo.toml`, or things like that. I think it's fine to review the existing revisions, and that will save me possibly having to resolve conflicts multiple times.

bjorn3 commented on compiler/rustc_feature/src/removed.rs:
```diff
@@ -155,9 +155,12 @@ declare_features! (
     (removed, panic_implementation, "1.28.0", Some(44489), None,
      Some("subsumed by `#[panic_handler]`")),
+    /// Allows using `#![plugin(myplugin)]`.
+    (removed, plugin, "1.74.0", Some(29597), None, // njn: version?
```
```suggestion
    (removed, plugin, "CURRENT_RUSTC_VERSION", Some(29597), None, // njn: version?
```
?

nnethercote replied:
Shouldn't it be `"1.75.0"`?

mati865 replied:
It will be replaced with proper version like in https://github.com/rust-lang/rust/pull/116407
This mechanism was introduced in https://github.com/rust-lang/rust/pull/100591

nnethercote replied:
I see, thanks for the explanation.

bjorn3 commented on compiler/rustc_interface/src/queries.rs:
```diff
@@ -148,12 +148,8 @@ impl<'tcx> Queries<'tcx> {
-                &pre_configured_attrs,
-            ));
+            let lint_store =
+                Lrc::new(passes::create_lint_store(sess, self.compiler.register_lints.as_deref()));
```
Doesn't need to be done right now, but maybe this can be moved into `build_session` now? I wouldn't be surprised if there are a couple of places where we are temporarily stashing lints just to wait on creating the lint store too. Those could be cleaned up too.

nnethercote replied:
Looking at this now... can you expand a little? I don't entirely understand the end point you are after. Would `GlobalCtxt` still hold the lint store?

nnethercote replied:
I have an in-progress change that moves the `lint_store` from `GlobalCtxt` to `Session`. Is that what you had in mind?

bjorn3 replied:
> Is that what you had in mind?

Yes. There is a place in the driver where a lint store is created before TyCtxt. With this change it would be possible to use the lint store stored in Session instead. Do you think the end result will be nicer than the current state?

nnethercote replied:
I did this in #117649.

bjorn3 commented:
LGTM with the feature removal version replaced by `CURRENT_RUSTC_VERSION`.

nnethercote commented:
@bors try

bors commented:
:hourglass: Trying commit 7e066318790caef8e093766e698d9d597cf91210 with merge c557db7a0ae416f6850160586b2d4c31eac75748...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7e066318790caef8e093766e698d9d597cf91210","merge_sha":"c557db7a0ae416f6850160586b2d4c31eac75748"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6699377273/job/18203394094) [(plain)](https://github.com/rust-lang/rust/commit/7e066318790caef8e093766e698d9d597cf91210/checks/18203394094/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.11s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
book/compiler-plugins.html:37: broken link - `unstable-book/language-features/plugin.html`
book/compiler-plugins.html:39: broken link - `unstable-book/language-features/plugin.html`
number of HTML files scanned: 34812
number of HTML redirects found: 10550
number of links checked: 2992556
number of links ignored due to external: 59231
```

</details>

nnethercote commented:
> ```
> book/compiler-plugins.html:37: broken link - `unstable-book/language-features/plugin.html`
> book/compiler-plugins.html:39: broken link - `unstable-book/language-features/plugin.html`
> ```

https://github.com/rust-lang/book/pull/3764 is my attempt to fix this.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6699442925/job/18205386046)
Build commit: c557db7a0ae416f6850160586b2d4c31eac75748 (`c557db7a0ae416f6850160586b2d4c31eac75748`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6699442925/job/18205386046"},"merge_sha":"c557db7a0ae416f6850160586b2d4c31eac75748"} -->

nnethercote committed:
Update rust-lang/book.

To 5b6c1ceaa62ecbd6caef08df39b33b3938e99deb, for this commit:

- Prepare for removal of compiler plugin support. (rust-lang/book#3764)

nnethercote commented:
@bors r=bjorn3

bors commented:
:pushpin: Commit 5c462a32bd31d19cacbd15786fb0fa40558eda91 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 5c462a32bd31d19cacbd15786fb0fa40558eda91 -->
<!-- homu: {"type":"Approved","sha":"5c462a32bd31d19cacbd15786fb0fa40558eda91","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6750523675/job/18353081229) [(plain)](https://github.com/rust-lang/rust/commit/5c462a32bd31d19cacbd15786fb0fa40558eda91/checks/18353081229/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit 5c462a32bd31d19cacbd15786fb0fa40558eda91 with merge 5020f7c3b8bd8bf57077389bca5c19f7911fde7a...
<!-- homu: {"type":"BuildStarted","head_sha":"5c462a32bd31d19cacbd15786fb0fa40558eda91","merge_sha":"5020f7c3b8bd8bf57077389bca5c19f7911fde7a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6750829863/job/18355724908)
Approved by: bjorn3
Pushing 5020f7c3b8bd8bf57077389bca5c19f7911fde7a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6750829863/job/18355724908"},"merge_sha":"5020f7c3b8bd8bf57077389bca5c19f7911fde7a"} -->

rust-timer commented:
Finished benchmarking commit (5020f7c3b8bd8bf57077389bca5c19f7911fde7a): [comparison URL](https://perf.rust-lang.org/compare.html?start=1bb6553b967b69eed9ed8147e78b4f65cfc48e11&end=5020f7c3b8bd8bf57077389bca5c19f7911fde7a&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1bb6553b967b69eed9ed8147e78b4f65cfc48e11&end=5020f7c3b8bd8bf57077389bca5c19f7911fde7a&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean   | range           | count |
|:----------------------------------:|:------:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -      | -               | 0     |
| Regressions ❌ <br /> (secondary)  | -      | -               | 0     |
| Improvements ✅ <br /> (primary)   | -10.1% | [-23.2%, -0.9%] | 8     |
| Improvements ✅ <br /> (secondary) | -      | -               | 0     |
| All ❌✅ (primary)                 | -10.1% | [-23.2%, -0.9%] | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1bb6553b967b69eed9ed8147e78b4f65cfc48e11&end=5020f7c3b8bd8bf57077389bca5c19f7911fde7a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.4%  | [1.1%, 1.8%]   | 3     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-3.5%, -0.7%] | 4     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.5%, -0.5%] | 5     |
| All ❌✅ (primary)                 | -2.1% | [-3.5%, -0.7%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1bb6553b967b69eed9ed8147e78b4f65cfc48e11&end=5020f7c3b8bd8bf57077389bca5c19f7911fde7a&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -               | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -9.3% | [-19.0%, -1.9%] | 7     |
| Improvements ✅ <br /> (secondary) | -     | -               | 0     |
| All ❌✅ (primary)                 | -9.3% | [-19.0%, -1.9%] | 7     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=1bb6553b967b69eed9ed8147e78b4f65cfc48e11&end=5020f7c3b8bd8bf57077389bca5c19f7911fde7a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 635.435s -> 634.085s (-0.21%)
**Artifact size**: 304.48 MiB -> 304.52 MiB (0.02%)

mati865 commented:
Perf improvements are mostly noise.
Previously there was fictional regression in PR that only modified docs: https://perf.rust-lang.org/compare.html?start=9c20ddd956426d577d77cb3f57a7db2227a3c6e9&end=1bb6553b967b69eed9ed8147e78b4f65cfc48e11&stat=instructions:u
After this PR the regression has disappeared.

nnethercote commented:
Postscript: @bjorn3's suggestion to move lint registration earlier was very useful. I did that in #117649, and it led to more clean-ups in #117993 and #118002. Combined with some recent related improvements I made in #117268, #117376, and #117475, and `rustc_driver_impl::run_compiler` and `rustc_interface::interface::run_compiler` are both looking quite a bit simpler and more readable than they used to. It's nice to see.
